### PR TITLE
Refactor #textEmphasis* features to avoid negative definitions (#838,…

### LINKED ISSUE #838: Typo at #textEmphasis-no-quoted-string.
`#textEmphasis-no-quoted-string` should be `#textEmphasis-no-color`

### LINKED ISSUE #841: Rework #textEmphasis* features.
Rework to avoid negative feature definitions, i.e., `-no-color` and `-no-quoted-string`.

### PR DESCRIPTION
**Pull request by skynavga**
…#841).

Closes #838.
Closes #841.

### CODE REVIEW

File: spec/ttml2.xml [hunk 1]
```diff
@@ -22003,64 +22003,53 @@ feature if it supports the <code>#textDecoration-under</code> feature as defined
...
+of transforming values of the <loc
 href="#style-attribute-textEmphasis"><att>tts:textEmphasis</att></loc>
-attribute, the value of which does not contain a <loc href="#style-value-emphasis-style">&lt;emphasis-style&gt;</loc> component that
-takes the form of a <loc href="#content-value-quoted-string">&lt;quoted-string&gt;</loc> value, or, if
-it contains said component value, then that component value is treated as if <code>auto</code> were specified.</p>
-<p>A TTML <loc href="#terms-presentation-processor">presentation processor</loc> supports the
-<code>#textEmphasis-no-color</code> feature if it implements presentation semantic support
-for all defined values of the <loc
+attribute which (1) do not contain an <loc href="#style-value-emphasis-color">&lt;emphasis-color&gt;</loc> component and
+(2) do not contain an <loc href="#style-value-emphasis-style">&lt;emphasis-style&gt;</loc> component that
```
Comment (nigelmegitt) [inline spec/ttml2.xml]: This formulation apparently prohibits `tts:textEmphasis` values from including either a color or quoted string emphasis-style component if this feature is supported, but I don't believe that is intended or desirable: the result we want to end up with is that those components may be included if the processor supports the related features.

Proposal: specify this in terms of which parts of the syntax the processor must minimally support rather than which parts of the syntax may be present. For example:

> A TTML transformation processor supports the `#textEmphasis-minimal` feature if it recognizes and is capable of transforming the following components of values of the `tts:textEmphasis` attribute:
> * `<emphasis-position>` and
> * `<emphasis-style>` excluding `<quoted-string>` values.

**Comment by nigelmegitt** [comment]
Additional review comment: you could also add an `extends="#textEmphasis-minimal"` attribute value on the profile entry for `#textEmphasis`.

File: spec/ttml2.xml (hunk 1 above)
Comment (skynavga) [reply spec/ttml2.xml]: @nigelmegitt you seem to have neglected the language "recognizes and is capable of transforming values ..."; therefore, it is already formulated in terms of what is supported; if color or quoted string values are present, but their transformation is not supported, then that usage is not supported, but it doesn't read on whether they may be present or not in content; so,  IMO, there is no problem here (except a typo I am just now fixing)

**Comment by skynavga** [comment]
@nigelmegitt fixed typo and missing extends

**Comment by nigelmegitt** [comment]
@skynavga ok they're good changes, what about my proposal at https://github.com/w3c/ttml2/pull/843#discussion_r196716082 ?

File: spec/ttml2.xml (hunk 1 above)
Comment (skynavga) [reply spec/ttml2.xml]: @nigelmegitt your reading is not supported by the language, which is consistent with other feature definitions of this form, e.g., `#condition-primary`, `#fontSize-isomorphic`, `#textOutline-unblurred`.

Comment (nigelmegitt) [reply spec/ttml2.xml]: @skynavga I disagree, or at least I'm seeing another potential reading that we should avoid by editing. The wording:

> ... it recognizes and is capable of transforming values of the tts:textEmphasis attribute which (1) do not contain an `<emphasis-color>` component and (2) do not contain an `<emphasis-style>` component that takes the form of a `<quoted-string>` value. 

suggests that the entire values which *do* contain either `<emphasis-color>` or (`<emphasis-style>` that is a `<quoted-string>`) or both do not need to be recognised or transformed at all by a processor that supports the feature.

I think it needs to say that the included components would be recognised or transformed, but say nothing about what happens to components that are not included in the feature.

In other words, how should a processor that supports the feature but not any of the other `#textEmphasis-x`features deal with, say, `tts:textEmphasis="auto blue before"`? I'm expecting it to recognise the `auto` and the `before` but do something unspecified with the `blue`, such as ignore it. But the current wording reads to me that the processor will fail to recognise or transform the *entire* value, even the bits it might be able to process.

Comment (skynavga) [reply spec/ttml2.xml]: >suggests that the entire values which do contain either `<emphasis-color>` or (`<emphasis-style>` that is a `<quoted-string>`) or both do not need to be recognised or transformed at all by a processor that supports the feature.

that's correct, and the desired semantics; we do not require (elsewhere) a processor to parse a construct containing unsupported syntax, though we could do so, but we have not done so yet, and would need to do the same in other similar contexts;

of course, a processor that knows about some possible construct but that only (semantically) supports part of it can be constructed, and that is a good approach for implementers; however, this doesn't protect one against new constructs not known at the time of implementation

>I think it needs to say that the included components would be recognised or transformed, but say nothing about what happens to components that are not included in the feature.

we do not specify this elsewhere in similar contexts; e.g., we don't say that the blur radius component must be parsed (or transformed) by a processor that supports `#textOutline-unblurred` but not `#textOutline-blurred`

>In other words, how should a processor that supports the feature but not any of the other `#textEmphasis-x` features deal with, say, `tts:textEmphasis="auto blue before"`? I'm expecting it to recognise the auto and the before but do something unspecified with the blue, such as ignore it. But the current wording reads to me that the processor will fail to recognise or transform the entire value, even the bits it might be able to process.

right now, we don't say either way (in general); robust implementations will generally parse known syntax the semantics of which are unsupported; I would not agree that the current language implies it will or must fail to recognize or transform; rather, it is silent on this matter, which is consistent with the other like features I enumerated

Comment (nigelmegitt) [reply spec/ttml2.xml]: Hmm, okay, I might be persuadable. Any other views? @palemieux perhaps?

**Comment by skynavga** [comment]
@nigelmegitt I think your proposal could be worded so as to be equivalent, but I think it does not add anything or improve the precision of the specification, so I am inclined to stay with the current (PR) text which is more consistent with the other uses I have cited;

on the general question of whether we should say more than nothing (about parsing known syntax that is not semantically supported), I wouldn't object to adding a note in the front matter (of E.1) that recommends parsing but ignoring unsupported value components

**Comment by nigelmegitt** [comment]
> I wouldn't object to adding a note in the front matter (of E.1) that recommends parsing but ignoring unsupported value components

I'd certainly support that.

**Comment by skynavga** [comment]
@nigelmegitt added such a note in https://github.com/w3c/ttml2/pull/843/commits/8426f59fa41b1524ca80efc9e97f1ec829f66e47

**Comment by nigelmegitt** [comment]
> added such a note in 8426f59

@skynavga 👍

File: spec/ttml2.xml (hunk 1 above)
Comment (nigelmegitt) [reply spec/ttml2.xml]: > we do not require (elsewhere) a processor to parse a construct containing unsupported syntax

Actually we don't say anything about it at all, so in that sense the text that triggered this thread is unusual. There are only two other `#xxx-minimal` features currently:

* `#rubyAlign-minimal` talks about the recognised values that the processor is capable of transforming.
* `#animate-minimal` uses the phrasing "if it recognizes and is capable of transforming the semantics of the `animate` element when used with the discrete and linear values of the `calcMode` attribute."

One difference here is that the syntax they reference does not have multiple components, whereas `tts:textEmphasis` does. It would be more consistent to relate this `#textEmphasis-minimal` feature to the supported values of specific components of the `tts:textEmphasis` attribute. Looked at this way, I'm back to preferring my proposal, since it does exactly that.

**Comment by skynavga** [comment]
@nigelmegitt since the current text is not broken and since it is consistent with the type of language expressed elsewhere, e.g., see `#condition-primary`, then I insist on following the editor's preference

**Comment by palemieux** [comment]
Built at https://rawgit.com/w3c/ttml2/issue-0841-text-emphasis-feature-build/index.html#feature-textEmphasis

**Comment by skynavga** [comment]
Merged early per WG resolution.

**Comment by skynavga** [comment]
Merged early per WG resolution.
